File: ios/brave-ios/Sources/Brave/Frontend/ParentalFreedom/Allow2CheckResult.swift
// Copyright 2024 The Brave Authors. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import Foundation

// MARK: - Allow2 Check Result

/// Response model from Allow2 check API
public struct Allow2CheckResult: Codable {
  /// Whether the requested activities are allowed
  public let allowed: Bool

  /// Activity-specific results keyed by activity ID
  public let activities: [String: ActivityResult]

  /// Day type information
  public let dayTypes: DayTypes?

  // MARK: - Computed Properties

  /// The minimum remaining seconds across all activities
  public var minimumRemainingSeconds: Int? {
    let remainingTimes = activities.values.compactMap { $0.remaining }
    return remainingTimes.min()
  }

  /// The soonest time block end across all activities
  public var soonestTimeBlockEnd: Date? {
    let endTimes = activities.values.compactMap { $0.timeblock?.endsDate }
    return endTimes.min()
  }

  /// Human-readable explanation of why access is blocked
  public var explanation: String {
    var reasons: [String] = []

    for (_, activity) in activities {
      if activity.banned {
        reasons.append("You are currently banned from \(activity.name).")
      } else if let timeblock = activity.timeblock, !timeblock.allowed {
        reasons.append("You cannot use \(activity.name) at this time.")
      } else if let remaining = activity.remaining, remaining <= 0 {
        reasons.append("You've used all your \(activity.name) time today.")
      }
    }

    return reasons.isEmpty ? "Access is restricted." : reasons.joined(separator: "\n")
  }

  /// Get result for a specific activity
  public func result(for activity: Allow2Activity) -> ActivityResult? {
    return activities[String(activity.rawValue)]
  }
}

// MARK: - Activity Result

/// Result for a specific activity type
public struct ActivityResult: Codable {
  /// Activity ID
  public let id: Int

  /// Activity name (e.g., "Internet", "Gaming")
  public let name: String

  /// Remaining time in seconds
  public let remaining: Int?

  /// Cache expiry timestamp
  public let expires: TimeInterval?

  /// Whether the child is banned from this activity
  public let banned: Bool

  /// Time block restrictions
  public let timeblock: TimeBlockResult?

  // MARK: - Computed Properties

  /// Whether this activity is currently allowed
  public var isAllowed: Bool {
    if banned { return false }
    if let timeblock = timeblock, !timeblock.allowed { return false }
    if let remaining = remaining, remaining <= 0 { return false }
    return true
  }

  /// Remaining time as TimeInterval
  public var remainingTimeInterval: TimeInterval? {
    guard let remaining = remaining else { return nil }
    return TimeInterval(remaining)
  }

  /// Cache expiry date
  public var expiresDate: Date? {
    guard let expires = expires else { return nil }
    return Date(timeIntervalSince1970: expires)
  }

  /// Formatted remaining time string (e.g., "15:00", "1:30:00")
  public var formattedRemainingTime: String? {
    guard let remaining = remaining, remaining > 0 else { return nil }

    let hours = remaining / 3600
    let minutes = (remaining % 3600) / 60
    let seconds = remaining % 60

    if hours > 0 {
      return String(format: "%d:%02d:%02d", hours, minutes, seconds)
    } else {
      return String(format: "%d:%02d", minutes, seconds)
    }
  }
}

// MARK: - Time Block Result

/// Time block restriction information
public struct TimeBlockResult: Codable {
  /// Whether access is allowed during current time block
  public let allowed: Bool

  /// Timestamp when this time block ends
  public let ends: TimeInterval?

  /// Time block end date
  public var endsDate: Date? {
    guard let ends = ends else { return nil }
    return Date(timeIntervalSince1970: ends)
  }

  /// Formatted time block end string
  public var formattedEndsTime: String? {
    guard let endsDate = endsDate else { return nil }
    let formatter = DateFormatter()
    formatter.dateStyle = .none
    formatter.timeStyle = .short
    return formatter.string(from: endsDate)
  }
}

// MARK: - Day Types

/// Information about day type (e.g., "School Night", "Weekend")
public struct DayTypes: Codable {
  /// Today's day type
  public let today: DayType?

  /// Tomorrow's day type
  public let tomorrow: DayType?
}

// MARK: - Day Type

/// Individual day type information
public struct DayType: Codable {
  /// Day type ID
  public let id: Int

  /// Day type name (e.g., "School Night", "Weekend")
  public let name: String
}

// MARK: - Request More Time

/// Request for additional time
public struct MoreTimeRequest: Codable {
  /// Activity to request time for
  public let activityId: Int

  /// Requested duration in minutes
  public let duration: Int

  /// Optional message to parent
  public let message: String?

  public init(activity: Allow2Activity, duration: Int, message: String? = nil) {
    self.activityId = activity.rawValue
    self.duration = duration
    self.message = message
  }
}

// MARK: - Request Duration Options

/// Predefined duration options for requesting more time
public enum MoreTimeDuration: Int, CaseIterable {
  case fifteenMinutes = 15
  case thirtyMinutes = 30
  case oneHour = 60
  case untilBedtime = -1  // Special value for "until bedtime"

  public var displayString: String {
    switch self {
    case .fifteenMinutes:
      return "15 minutes"
    case .thirtyMinutes:
      return "30 minutes"
    case .oneHour:
      return "1 hour"
    case .untilBedtime:
      return "Until bedtime"
    }
  }
}

File: android/java/org/chromium/chrome/browser/allow2/models/Allow2CheckResult.kt
/* Copyright (c) 2024 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at https://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser.allow2.models

import org.json.JSONObject

/**
 * Data class representing the result of an Allow2 check API call.
 */
data class Allow2CheckResult(
    val allowed: Boolean,
    val activities: Map<Int, ActivityResult>,
    val todayDayType: DayType?,
    val tomorrowDayType: DayType?,
    val expiresAt: Long
) {
    /**
     * Get the primary activity result (Internet).
     */
    fun getPrimaryActivityResult(): ActivityResult? {
        return activities[Allow2Activity.INTERNET.id]
    }

    /**
     * Get the remaining time in seconds for the primary activity.
     */
    fun getRemainingSeconds(): Int {
        return getPrimaryActivityResult()?.remaining ?: 0
    }

    /**
     * Check if the cached result is still valid.
     */
    fun isExpired(): Boolean {
        return System.currentTimeMillis() / 1000 > expiresAt
    }

    /**
     * Get a user-friendly block reason.
     */
    fun getBlockReason(): String {
        val activity = getPrimaryActivityResult()
        return when {
            activity == null -> "Access restricted"
            activity.banned -> "Access has been blocked"
            activity.timeblock != null && !activity.timeblock.allowed ->
                "Not allowed during ${todayDayType?.name ?: "this time"}"
            activity.remaining <= 0 -> "Time limit reached for today"
            else -> "Access restricted"
        }
    }

    /**
     * Convert to JSON for caching.
     */
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("allowed", allowed)
            put("expiresAt", expiresAt)

            val activitiesJson = JSONObject()
            activities.forEach { (id, result) ->
                activitiesJson.put(id.toString(), result.toJson())
            }
            put("activities", activitiesJson)

            todayDayType?.let { put("todayDayType", it.toJson()) }
            tomorrowDayType?.let { put("tomorrowDayType", it.toJson()) }
        }
    }

    companion object {
        /**
         * Parse check result from API response.
         */
        fun fromJson(json: JSONObject): Allow2CheckResult {
            val activitiesJson = json.optJSONObject("activities")
            val activities = mutableMapOf<Int, ActivityResult>()

            if (activitiesJson != null) {
                val keys = activitiesJson.keys()
                while (keys.hasNext()) {
                    val key = keys.next()
                    val activityId = key.toIntOrNull()
                    if (activityId != null) {
                        val activityJson = activitiesJson.optJSONObject(key)
                        if (activityJson != null) {
                            activities[activityId] = ActivityResult.fromJson(activityJson)
                        }
                    }
                }
            }

            val dayTypesJson = json.optJSONObject("dayTypes")
            val todayDayType = dayTypesJson?.optJSONObject("today")?.let { DayType.fromJson(it) }
            val tomorrowDayType = dayTypesJson?.optJSONObject("tomorrow")?.let { DayType.fromJson(it) }

            // Calculate expiry from the activity expires field, or default to 60 seconds
            var expiresAt = System.currentTimeMillis() / 1000 + 60
            activities.values.firstOrNull()?.let { activity ->
                if (activity.expires > 0) {
                    expiresAt = activity.expires
                }
            }

            return Allow2CheckResult(
                allowed = json.optBoolean("allowed", true),
                activities = activities,
                todayDayType = todayDayType,
                tomorrowDayType = tomorrowDayType,
                expiresAt = expiresAt
            )
        }

        /**
         * Create a default "allowed" result.
         */
        fun allowed(): Allow2CheckResult {
            return Allow2CheckResult(
                allowed = true,
                activities = emptyMap(),
                todayDayType = null,
                tomorrowDayType = null,
                expiresAt = System.currentTimeMillis() / 1000 + 60
            )
        }

        /**
         * Create a default "blocked" result.
         */
        fun blocked(reason: String = "Access restricted"): Allow2CheckResult {
            return Allow2CheckResult(
                allowed = false,
                activities = mapOf(
                    Allow2Activity.INTERNET.id to ActivityResult(
                        id = Allow2Activity.INTERNET.id,
                        name = "Internet",
                        remaining = 0,
                        expires = 0,
                        banned = true,
                        timeblock = null
                    )
                ),
                todayDayType = null,
                tomorrowDayType = null,
                expiresAt = System.currentTimeMillis() / 1000 + 60
            )
        }
    }
}

/**
 * Result for a specific activity.
 */
data class ActivityResult(
    val id: Int,
    val name: String,
    val remaining: Int,
    val expires: Long,
    val banned: Boolean,
    val timeblock: TimeBlock?
) {
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("id", id)
            put("name", name)
            put("remaining", remaining)
            put("expires", expires)
            put("banned", banned)
            timeblock?.let { put("timeblock", it.toJson()) }
        }
    }

    companion object {
        fun fromJson(json: JSONObject): ActivityResult {
            val timeblockJson = json.optJSONObject("timeblock")
            return ActivityResult(
                id = json.optInt("id", 0),
                name = json.optString("name", ""),
                remaining = json.optInt("remaining", 0),
                expires = json.optLong("expires", 0),
                banned = json.optBoolean("banned", false),
                timeblock = timeblockJson?.let { TimeBlock.fromJson(it) }
            )
        }
    }
}

/**
 * Time block information (scheduled restrictions).
 */
data class TimeBlock(
    val allowed: Boolean,
    val ends: Long
) {
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("allowed", allowed)
            put("ends", ends)
        }
    }

    companion object {
        fun fromJson(json: JSONObject): TimeBlock {
            return TimeBlock(
                allowed = json.optBoolean("allowed", true),
                ends = json.optLong("ends", 0)
            )
        }
    }
}

/**
 * Day type information (e.g., "School Night", "Weekend").
 */
data class DayType(
    val id: Int,
    val name: String
) {
    fun toJson(): JSONObject {
        return JSONObject().apply {
            put("id", id)
            put("name", name)
        }
    }

    companion object {
        fun fromJson(json: JSONObject): DayType {
            return DayType(
                id = json.optInt("id", 0),
                name = json.optString("name", "")
            )
        }
    }
}
